### PR TITLE
User, Group, and Billing Project Components and tests

### PR DESCRIPTION
**Pull request by jmthibault79**
DSDEEPB-2802
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [ ] Tell  that the PR exists if he wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [x] Comment / review / update cycle:
  - Rest of team may comments on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter updates documentation as needed
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [x]  sign off
- [x] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Squash commits, rebase if necessary
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Merge to develop 
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: Check configuration files in Jenkins in case they need changes
- [ ] **Submitter**: Verify swagger UI on dev server still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsBillingProjectComponent.scala [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+    private def fromStrings = (name: String, bucketUrl: String) => RawlsBillingProjectShort(RawlsBillingProjectName(name), bucketUrl)
+    private def toStrings = (project: RawlsBillingProjectShort) => Option(project.projectName.value, project.cromwellAuthBucketUrl.value)
+
+    def * = (projectName, cromwellAuthBucketUrl) <> (fromStrings.tupled, toStrings)
+  }
+
+  val rawlsBillingProjectShortQuery = TableQuery[RawlsBillingProjectTable]
+
+  def rawlsBillingProjectByNameQuery(name: String) = Compiled {
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsBillingProjectComponent.scala]: This could be private and/or not compiled.  What do you think?

Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsBillingProjectComponent.scala]: I don't have a strong opinion on this. Maybe we could keep it as is and once we better learn about it, we can standardize across all queries.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsBillingProjectComponent.scala [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+import org.broadinstitute.dsde.rawls.model._
+
+trait RawlsBillingProjectComponent {
+  this: DriverComponent =>
+
+  import driver.api._
+
+  class RawlsBillingProjectTable(tag: Tag) extends Table[RawlsBillingProjectShort](tag, "BILLING_PROJECT") {
+    def projectName = column[String]("projectName", O.PrimaryKey)
```
Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsBillingProjectComponent.scala]: From what I've seen, strings are stored as text in the db by default. Since a lot of these things won't be arbitrarily long, do we want to be more efficient and store them as varchars?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsGroupComponent.scala [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import org.broadinstitute.dsde.rawls.model.{RawlsGroupShort, RawlsGroupEmail, RawlsGroupName}
+
+trait RawlsGroupComponent {
+  this: DriverComponent =>
+
+  import driver.api._
+
+  class RawlsGroupTable(tag: Tag) extends Table[RawlsGroupShort](tag, "GROUP") {
+    def groupName = column[String]("groupName", O.PrimaryKey)
+    def groupEmail = column[String]("groupEmail")
```
Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsGroupComponent.scala]: In the schema google doc, groupEmail is listed as the proposed primary key. Is this change on purpose?

**Comment by MatthewBemis** [comment]
A couple of comments. After they're discussed/addressed, :+1:. I'm also assuming that user membership and subgroup membership tables will be covered in another ticket.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsGroupComponent.scala (hunk 3 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsGroupComponent.scala]: Yes.  `RawlsUserRef` uses `groupName`, so I changed it for consistency with the existing code.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsBillingProjectComponent.scala (hunk 2 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsBillingProjectComponent.scala]: Does it make a big difference?  VARCHAR isn't a standard Slick type, so if it's not clear what the performance improvement would be, I'd prefer to keep it more generic.

**Comment by jmthibault79** [comment]
Yes, I'll handle Project Membership, User Membership, and Subgroup Membership in a different PR.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsBillingProjectComponent.scala (hunk 2 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsBillingProjectComponent.scala]: I expect that we will create the DDL separately via liquibase so we can specify varchar there.

**Comment by dvoet** [comment]
why aren't there case classes for each of the tables?

**Comment by jmthibault79** [comment]
Because there were existing "Short" case classes that corresponded to the tables.  But I like your `Record` approach better and I will update accordingly.

**Comment by jmthibault79** [comment]
Updated to use the Record approach.  Please re-review.

**Comment by dvoet** [comment]
:+1:
